scripts/wallet.js

```diff
@@ -1073,23 +1030,87 @@ export class Wallet {
     }
 
     /**
-     * Finalize Transaction. To be called after it's signed and sent to the network, if successful
-     * @param {Transaction} transaction
+     * Adds a transaction to the mempool. To be called after it's signed and sent to the network, if successful
+     * @param {import('./transaction.js').Transaction} transaction
      */
-    finalizeTransaction(transaction) {
+    async addTransaction(transaction, skipDatabase = false) {
+        this.#mempool.addTransaction(transaction);
+        let i = 0;
+        for (const out of transaction.vout) {
+            this.updateHighestUsedIndex(out);
+            const status = this.getScriptType(out.script);
+            if (status & OutpointState.OURS) {
+                this.#mempool.setOutpointStatus(
+                    new COutpoint({
+                        txid: transaction.txid,
+                        n: i,
+                    }),
+                    status
+                );
+            }
+            i++;
+        }
+
         if (transaction.hasShieldData) {
             wallet.#shield.finalizeTransaction(transaction.txid);
         }
 
-        mempool.updateMempool(transaction);
-        mempool.setBalance();
+        if (skipDatabase) {
+            const db = await Database.getInstance();
+            await db.storeTx(transaction);
+        }
+    }
+
+    /**
+     * @returns {UTXO[]} Any UTXO that has value of
+     * exactly `cChainParams.current.collateralInSats`
+     */
+    getMasternodeUTXOs() {
+        const collateralValue = cChainParams.current.collateralInSats;
+        return this.#mempool
+            .getUTXOs({
+                filter: OutpointState.SPENT | OutpointState.IMMATURE,
+                requirement: OutpointState.P2PKH,
+            })
+            .filter((u) => u.value === collateralValue);
+    }
+
+    /**
+     * @returns {import('./transaction.js').Transaction[]} a list of all transactions
+     */
+    getTransactions() {
+        return this.#mempool.getTransactions();
+    }
+
+    get balance() {
+        return this.#mempool.balance;
+    }
+
+    get immatureBalance() {
+        return this.#mempool.immatureBalance;
+    }
+
+    get coldBalance() {
+        return this.#mempool.coldBalance;
+    }
+
+    async loadFromDisk() {
+        const db = await Database.getInstance();
+        if ((await db.getAccount())?.publicKey !== this.getKeyToExport()) {
+            await db.removeAllTxs();
+            return;
+        }
+        const txs = await db.getTxs();
+        for (const tx of txs) {
+            this.addTransaction(tx, true);
```

Comment:
you should order the transaction by blockheight  before calling this.addTransaction

Reply:
I believe this is not necessary unless we want to remove `sort` in the activity

Reply:
Resolved on discord